force-app/main/default/classes/OpportunityHelper.cls

```diff
@@ -0,0 +1,240 @@
+    ];
+
+    // Filter opportunities without a primary contact and not deleted
+    for (Opportunity opp : newOppList) {
+        if (opp.Primary_Contact__c == null && !opp.IsDeleted) {
+
+            // Clone the existing record to maintain original data and avoid modifying the original list directly
+            Opportunity updatedOpp = opp.clone(true, true, false); 
+            
+            updatedOpp.Primary_Contact__c = vpSalesContacts.get(0).Id;  // Set the Primary_Contact__c field
+
+            oppToUpdate.add(updatedOpp);  // Add the modified clone to the oppToUpdate list
+
+        }
+
+    }
+
+    // Optional: Perform the update if needed
+    try {
+        update oppToUpdate;
+        System.debug('Opportunities updated successfully: ' + oppToUpdate.size());
+    } catch (DmlException e) {
+        System.debug('Error updating Opportunities: ' + e.getMessage());
+    }
+
+}
+
+/*
+    setDefaultType:
+    - Assigns a Type for new Opportunities.
+    - Only updates the Opportunities that doesn't have an assigned Type.
+    */  
+    public static void setDefaultType(List<Opportunity> newOppList) {
+
+        // Set default Type for new Opportunities
+        for(Opportunity opp : newOppList) {
+            if (opp.Type == null){
+                    opp.Type = 'New Customer';
+            }        
+        }
+    }
+
+    /*
+    createNewTask:
+    - Create a new Task for newly inserted Opportunities
+    */  
+    public static void createNewTask(List<Opportunity> newOppList) {
+        // Check for Task read and create permissions
+        if (!Schema.sObjectType.Task.isAccessible()
+            && !Schema.sObjectType.Task.isCreateable()) {
+            throw new OtherException('Current user does not have permission to read and create Task records.');
+        }
+
+        List<Task> newTaskList = new List<Task>();
+
+        for (Opportunity opp : newOppList){
+            Task tsk = new Task();
+            tsk.Subject = 'Call Primary Contact';
+            tsk.WhatId = opp.Id;
+            tsk.WhoId = opp.Primary_Contact__c;
+            tsk.OwnerId = opp.OwnerId;
+            tsk.ActivityDate = Date.today().addDays(3);
+            newTaskList.add(tsk);
+        }
+        
+        insert newTaskList;
+
+    }
+
+    /*
+    appendStageChanges:
+    - Append Stage changes in Opportunity Description
+    */  
+    public static void appendStagenameChanges(List<Opportunity> newOppList) {
+        // Check for Opportunity read and update permissions
+        if (!Schema.sObjectType.Opportunity.isAccessible()
+            && !Schema.sObjectType.Opportunity.isUpdateable()) {
+            throw new OtherException('Current user does not have permission to read and update Opportunity records.');
+        }
+
+        List<Opportunity> updateOppList = new List<Opportunity>();
+
+        for (Opportunity opp : newOppList){
+
+            if (opp.StageName != null){
+
```

Comment:
Try to check to see if the stage has changed. Compare the old opp to the new opp, if there was a change only then make the update